app/bundles/EmailBundle/Helper/MailerDsnConvertor.php

```diff
@@ -0,0 +1,157 @@
+            $parameters['mailer_host']      = $dsn->getHost();
+            $parameters['mailer_port']      = $dsn->getPort();
+            $parameters['mailer_user']      = $dsn->getUser();
+            $parameters['mailer_password']  = $dsn->getPassword();
+
+            foreach ($dsn->getOptions() as $option => $value) {
+                $parameters['mailer_option_'.$option] = $value;
+            }
+        } catch (\InvalidArgumentException $e) {
+            $dsn                                     = Dsn::fromString('smtp://localhost:25');
+            $parameters['mailer_dsn']                = $dsnString;
+            $parameters['mailer_transport']          = $dsn->getScheme();
+            $parameters['mailer_host']               = $dsn->getHost();
+            $parameters['mailer_port']               = $dsn->getPort();
+            $parameters['mailer_user']               = $dsn->getUser();
+            $parameters['mailer_password']           = $dsn->getPassword();
+            $parameters['mailer_option_invalid_dsn'] = true;
+        }
+
+        return $parameters;
+    }
+
+    /**
+     * Convert the DSN options to a DSN string.
+     *
+     * @param array<string> $parameters     The array of DSN options
+     * @param array<string> $convertorClass The class to use for converting the DSN options
+     *
+     * @return string The DSN string
+     */
+    public static function convertArrayToDsnString(array $parameters, array $convertorClass = []): string
+    {
+        /**
+         * We need to make sure that we have
+         * both the user name and password before storing.
+         * If the username is empty we need to flush the old password.
+         */
+        $password                      = (!empty($parameters['mailer_user'])) ? self::getPassword($parameters) : null;
+        $user                          = (!empty($parameters['mailer_user'])) ? $parameters['mailer_user'] : null;
+        $parameters['mailer_password'] = $password;
+
+        if ('smtp' === $parameters['mailer_transport']) {
+            $host = self::getDefaultHost($parameters);
+            if (empty($host)) {
+                return '';
+            }
+
+            $options = [];
+            foreach (self::SUPPORTED_OPTIONS as $option => $parameterName) {
+                if (array_key_exists($parameterName, $parameters) && !empty($parameters[$parameterName])) {
+                    $options[$option] = $parameters[$parameterName];
+                }
+            }
+
+            return DsnGenerator::getDsnString(
+                new Dsn(
+                    $parameters['mailer_transport'],
+                    $host,
+                    $user,
+                    $password,
+                    self::getPort($parameters),
+                    $options
+                )
+            );
+        } else {
+            $class_name = $convertorClass[$parameters['mailer_transport']];
```

Comment:
@escopecz I am not fan of how I wrote this code, do you have a better suggestion?

Reply:
As I wrote above, I'd go only 1 direction and that is to convert all email params into DSN as a 1 time migration. So this wouldn't be needed at all.